pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,410 @@
+        decomposition.
+
+    """
+    if any(isinstance(m, VarianceMP) for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of variance with hadamard test gradient is not implemented yet."
+        )
+    if any(isinstance(m, (StateMP, VnEntropyMP, MutualInfoMP)) for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of circuits that return the state is not supported."
+        )
+
+    if argnum is None and not tape.trainable_params:
+        return _no_trainable_grad_new(tape, shots)
+
+    gradient_analysis(tape, grad_fn=hadamard_grad)
+    method = "analytic"
+    diff_methods = grad_method_validation(method, tape)
+
+    if all(g == "0" for g in diff_methods):
+        return _all_zero_grad_new(tape, shots)
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    argnum = [i for i, dm in method_map.items() if dm == "A"]
+
+    if device_wires and len(tape.wires) == len(device_wires):
+        raise qml.QuantumFunctionError("The device has no free wire for the auxiliary wire.")
+
+    # Get default for aux_wire
+    if aux_wire is None:
+        aux_wire = _get_aux_wire(aux_wire, tape, device_wires)
+    elif aux_wire[0] in tape.wires:
+        raise qml.QuantumFunctionError("The auxiliary wire is already used.")
+    elif aux_wire[0] not in device_wires:
+        raise qml.QuantumFunctionError("The requested aux_wire does not exist on the used device.")
+
+    g_tapes, processing_fn = _expval_hadamard_grad(tape, argnum, aux_wire)
+
+    return g_tapes, processing_fn
+
+
+def _expval_hadamard_grad(tape, argnum, aux_wire):
+    r"""Compute the Hadamard test gradient of a tape that returns an expectation value (probabilities are expectations
+    values) with respect to a given set of all trainable gate parameters.
+    The auxiliary wire is the wire which is used to apply the Hadamard gates and controlled gates.
+    """
+    # pylint: disable=too-many-statements
+    argnums = argnum or tape.trainable_params
+    g_tapes = []
+    coeffs = []
+
+    gradient_data = []
+    measurements_probs = [
+        idx
+        for idx, m in enumerate(tape.measurements)
+        if isinstance(m, qml.measurements.ProbabilityMP)
+    ]
+    for id_argnum, _ in enumerate(tape.trainable_params):
+        if id_argnum not in argnums:
+            # parameter has zero gradient
+            gradient_data.append(0)
+            continue
+
+        trainable_op, idx, p_idx = tape.get_operation(id_argnum, return_op_index=True)
+
+        ops_to_trainable_op = tape.operations[: idx + 1]
+        ops_after_trainable_op = tape.operations[idx + 1 :]
+
+        # Get a generator and coefficients
+        sub_coeffs, generators = _get_generators(trainable_op)
+        coeffs.extend(sub_coeffs)
+
+        num_tape = 0
+
+        for gen in generators:
+            if isinstance(trainable_op, qml.Rot):
```

Comment:
Where could one find information on how `Rot` is handled? Is there a reference for this?

Reply:
Basically you decompose `Rot=RzRyRz` and depending on which parameter is trainable you have three cases

1. First parameter `Rz PauliZ Ry Rz` becomes `ctrl-PauliZ Rot`
2. Third parameter `Rz Ry Rz PauliZ` becomes `Rot ctrl-PauliZ`
3. Secon parameter `Rz Ry PauliY Rz` becomes `Rz Ry ctrl-PauliY Rz` and can be simplified `Rot Rz Rx(pi/2) ctrl-PauliZ Rx(pi/2) Rz`

Reply:
Should this information, in one way or another, be in a comment?

Reply:
I made a suggestion below, but no blocker.